### DISCUSSION TITLE
Middleware: as today, we are still forced to run on edge?

### DISCUSSION THREAD

**Discussion by FrancescoSaverioZuppichini** [vercel/next.js#72085] [Help] [answered]
### Summary

I am wondering if middleware still are forced to run on edge, if this is true, why is still the case?

Thanks a lot,

Fra

**Comment by timneutkens** [comment]
https://github.com/vercel/next.js/discussions/71727

**Reply by FrancescoSaverioZuppichini** [reply]
Thanks a lot, so the answer is not yet! Thanks unfortunate, I'll have to switch framework at my company since I cannot use auth.js to my understanding and self host

**Reply by icyJoseph** [reply]
If you are self-hosting, then the middleware will run under regular Node.js, but when compiling, you'd be expected to follow the Edge runtime rules. 

What is the issue you have?

**Comment by leerob** [comment, marked as answer]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
